PurchasesCoreSwift/Public/PurchaserInfo.swift

```diff
@@ -0,0 +1,317 @@
+        let allPurchases: [String: Any]
+        let expirationDatesByProduct: [String: Date?]
+        let purchaseDatesByProduct: [String: Date?]
+
+        init?(subscriberData: [String: Any]) {
+            let subscriptions = subscriberData["subscriptions"] as? [String: [String: Any]] ?? [String: [String: Any]]()
+
+            // Metadata
+            self.originalApplicationVersion = subscriberData["original_application_version"] as? String ?? nil
+
+            if let originalPurchaseDateString = subscriberData["original_purchase_date"] as? String {
+                self.originalPurchaseDate = PurchaserInfo.parseDate(date: originalPurchaseDateString, withDateFormatter: DateFormatter.dateFormatter)
+            } else {
+                self.originalPurchaseDate = nil
+            }
+
+            guard let firstSeenDateString = subscriberData["first_seen"] as? String,
+                  let firstSeenDate = PurchaserInfo.parseDate(date: firstSeenDateString, withDateFormatter: DateFormatter.dateFormatter) else {
+                return nil
+            }
+            self.firstSeen = firstSeenDate
+
+            guard let originalAppUserIdString = subscriberData["original_app_user_id"] as? String else {
+                return nil
+            }
+            self.originalAppUserId = originalAppUserIdString
+
+            if let managementUrlString = subscriberData["management_url"] as? String {
+                self.managementURL = parseURL(url: managementUrlString)
+            } else {
+                self.managementURL = nil
+            }
+
+            // Purchases and entitlements
+            self.nonSubscriptions = subscriberData["non_subscriptions"] as? [String: [[String: Any]]] ?? [String: [[String: Any]]]()
+            self.entitlements = subscriberData["entitlements"] as? [String: Any] ?? [String: Any]()
+            self.nonSubscriptionTransactions = TransactionsFactory().nonSubscriptionTransactions(withSubscriptionsData: nonSubscriptions, dateFormatter: DateFormatter.dateFormatter)
+
+            let latestNonSubscriptionTransactionsByProductId = [String: Any](uniqueKeysWithValues: nonSubscriptions.map { productId, transactionsArray in
+                (productId, transactionsArray.last)
+            })
+
+            self.allPurchases = latestNonSubscriptionTransactionsByProductId.merging(subscriptions) { (current, _) in current }
+
+            self.expirationDatesByProduct = PurchaserInfo.parseExpirationDate(expirationDates: subscriptions)
+            self.purchaseDatesByProduct = PurchaserInfo.parsePurchaseDate(purchaseDates: allPurchases)
+        }
+    }
+
+}
+
+private extension PurchaserInfo {
+
+    func activeKeys(dates: [String: Date?]) -> Set<String> {
+        return Set<String>(dates.keys.filter({
+            if let nonNullDate = dates[$0] as? Date {
+                return isAfterReferenceDate(date: nonNullDate)
+            }
+            return true
+        }))
```

Comment:
is the `Set<String>` here needed? 
would doing 
```swift
return dates.keys.filter {
    guard let nonNullDate = dates[$0] as? Date else { return true }
    return isAfterReferenceDate(date: nonNullDate)
}
```
be equivalent?

Reply:
`keys` returns an Array, so it looks like I do have to construct a set from it...but i like your filtering code better 👍